scan/context.go

```diff
@@ -88,35 +87,40 @@ func (s *Scanner) getContextFromGitURL(gitURL string) (contextDir string, err er
 }
 
 func (s *Scanner) getContextFromURL(remoteURL string) (err error) {
-	response, err := getWithStatusError(remoteURL)
-	if err != nil {
-		return errors.Wrapf(err, "unable to download remote context from %s", remoteURL)
-	}
+	attempt := 0
+	for attempt < maxRetries {
+		var response *http.Response
+		response, err = getWithStatusError(remoteURL)
+		if err != nil {
+			return errors.Wrapf(err, "unable to download remote context from %s", remoteURL)
+		}
 
-	// TODO: revamp streaming, for now just pipe to buf and discard it.
-	var buf bytes.Buffer
-	progressOutput := streamformatter.NewProgressOutput(&buf)
-	r := progress.NewProgressReader(response.Body, progressOutput, response.ContentLength, "", "Downloading build context")
-	defer func(response *http.Response) {
-		_ = response.Body.Close()
-	}(response)
+		fmt.Printf("Read context with status code %d\n", response.StatusCode)
+		fmt.Printf("Read context of %d bytes\n", response.ContentLength)
 
-	err = s.getContextFromReader(r)
-	return err
-}
+		// TODO: revamp output streaming, for now just discard it
+		progressOutput := streamformatter.NewProgressOutput(io.Discard)
 
-func (s *Scanner) getContextFromReader(r io.Reader) (err error) {
-	buf := bufio.NewReader(r)
-	var magic []byte
+		var rc io.ReadCloser = progress.NewProgressReader(response.Body, progressOutput, response.ContentLength, "", "Downloading build context")
+		defer rc.Close()
 
-	// note: (sam) read 2048 magic bytes to accomodate Extended Pax headers.
-	magic, err = buf.Peek(archiveHeaderSize * 4)
-	if err != nil && err != io.EOF {
-		return errors.Wrap(err, "failed to peek context header")
+		err = s.getContextFromReader(rc)
+		if err == nil {
+			return nil
+		}
+
+		attempt++
+		time.Sleep(util.GetExponentialBackoff(attempt))
 	}
 
-	if dockerbuild.IsArchive(magic) {
```

Comment:
Would like to discuss is it necessary to peek the magic bytes and check whether the content is archive? If its content is not valid, we can let `archive.Untar()` return the error.

Reply:
The original code will skip untar and return. The pr changes the logic. Do you think it's related to the netowrk issue? If not, I suggest we keep the original logic in this round.

Reply:
Sure, rolled back the logic to checking archive before doing untar. I don't think it's related to the issue, just trying to make the streaming a little bit simpler.